src/cli/npc.rs

```diff
@@ -1,41 +1,92 @@
+    pub msg: String,
+}
+
+impl Display for CliError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{}", self.msg)
+    }
+}
+
+#[derive(Serialize)]
+enum CliResult {
+    Full(NetState),
+    Ifaces(Vec<Iface>),
+    Routes(Vec<Route>),
+    RouteRules(Vec<RouteRule>),
+    Error(CliError),
+    NisporError(NisporError),
+}
 
 fn main() {
-    let argv: Vec<String> = args().collect();
-    match NetState::retrieve() {
-        Ok(state) => {
-            if argv.len() > 1 {
-                match &state.ifaces.get(&argv[1]) {
-                    Some(iface) => println!(
-                        "{}",
-                        serde_json::to_string_pretty(iface).unwrap()
-                    ),
-                    None => {
-                        if argv[1] == "route" {
-                            println!(
-                                "{}",
-                                serde_json::to_string_pretty(&state.routes)
-                                    .unwrap()
-                            );
-                        } else if argv[1] == "rule" {
-                            println!(
-                                "{}",
-                                serde_json::to_string_pretty(&state.rules)
-                                    .unwrap()
-                            );
-                        } else {
-                            eprintln!("Interface '{}' not found", argv[1]);
-                        }
-                    }
+    let matches = clap_app!(npc =>
+        (version: crate_version!())
+        (author: crate_authors!())
+        (about: "Nispor CLI")
+        (@arg ifname: [INTERFACE_NAME] "interface name")
+        (@subcommand route =>
+            (about: "Show routes")
+        )
+        (@subcommand rule =>
+            (about: "Show routes rules")
+        )
+    )
+    .get_matches();
+
+    let result = match NetState::retrieve() {
+        Ok(mut state) => {
+            if let Some(ifname) = matches.value_of("ifname") {
+                if let Some(iface) = state.ifaces.remove(ifname) {
+                    CliResult::Ifaces(vec![iface])
+                } else {
+                    CliResult::Error(CliError {
+                        msg: format!("Interface '{}' not found", ifname),
+                    })
                 }
+            } else if let Some(_) = matches.subcommand_matches("route") {
+                CliResult::Routes(state.routes)
+            } else if let Some(_) = matches.subcommand_matches("rule") {
+                CliResult::RouteRules(state.rules)
             } else {
-                println!("{}", serde_json::to_string_pretty(&state).unwrap());
+                /* Show everything if no cmdline arg has been supplied */
+                CliResult::Full(state)
             }
         }
-        Err(e) => {
-            eprintln!("{}", e);
+        Err(e) => CliResult::NisporError(e),
+    };
+
+    match result {
+        CliResult::Full(netstate) => {
```

Comment:
If can write a macro for this it'd be great, I'm not really that skilled in rust ;)

Reply:
```rust
match result {
    CliResult::Error(e) | CliResult::NisporError(e) => something,
    _ => println!()
}
```

Reply:
Unfortunately that does not work, all binding must have the same type in all alternatives.